src/base/searchengine.h

```diff
@@ -30,18 +30,135 @@
 #ifndef SEARCHENGINE_H
 #define SEARCHENGINE_H
 
+#include <array>
+
 #include <QObject>
 #include <QHash>
+#include <QMetaType>
 #include <QStringList>
 #include <QList>
 
 class QProcess;
 class QTimer;
 
+template <class T, std::size_t N>
+class Version
+{
+public:
+    typedef T ComponentType;
+    typedef Version<T,N> ThisType;
+
+    Version()
+    {
+        m_components.fill(T());
+    }
+
+    template <class ...Other>
+    Version(Other... components)
+        : m_components {{components...}}
+    {
+    }
+
+    Version(const QString &version)
+        : Version {version.split(QLatin1Char('.'))}
+    {
+    }
+
+    Version(const QByteArray &version)
+        : Version {version.split('.')}
+    {
+    }
+
+    ComponentType majorNumber() const
+    {
+        static_assert(N >= 1, "Array size is too small");
+        return (*this)[0];
+    }
+
+    ComponentType minorNumber() const
+    {
+        static_assert(N >= 2, "Array size is too small");
+        return (*this)[1];
+    }
+
+    ComponentType revisionNumber() const
+    {
+        static_assert(N >= 2, "Array size is too small");
+        return (*this)[2];
+    }
+
+    ComponentType patchNumber() const
+    {
+        static_assert(N >= 3, "Array size is too small");
+        return (*this)[3];
+    }
+
+    ComponentType operator[](std::size_t i) const
+    {
+        return m_components.at(i);
+    }
+
+    operator QString() const
+    {
+        QString res = QString::number((*this)[0]);
+        for (std::size_t i = 1; i < N; ++i) {
+            res += QLatin1Char('.') + QString::number((*this)[i]);
+        }
+        return res;
+    }
+
+    bool operator==(const ThisType &other) const
+    {
+        return m_components == other.m_components;
+    }
+
+    bool operator!=(const ThisType &other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator<(const ThisType &other) const
+    {
+        for (std::size_t i = 0; i < N - 1; ++i) {
+            if ((*this)[i] > other[i])
+                return false;
+        }
+        return (*this)[N-1] < other[N-1];
+    }
+
+    bool operator>(const ThisType &other) const
+    {
+        for (std::size_t i = 0; i < N - 1; ++i) {
+            if ((*this)[i] < other[i])
+                return false;
+        }
+        return (*this)[N-1] > other[N-1];
```

Comment:
BTW, I forgot to reply to a previous comment. I use this->operator[] in order to encapsulate the implementation (`std::array`) and catch potential out of range errors.